manticore/native/cpu/abstractcpu.py

```diff
@@ -950,19 +950,21 @@ def execute(self):
         """
         Decode, and execute one instruction pointed by register PC
         """
-        if issymbolic(self.PC):
+        curpc = self.PC
+        if issymbolic(curpc):
             raise ConcretizeRegister(self, "PC", policy="ALL")
-        if not self.memory.access_ok(self.PC, "x"):
-            raise InvalidMemoryAccess(self.PC, "x")
+        if not self.memory.access_ok(curpc, "x"):
+            raise InvalidMemoryAccess(curpc, "x")
 
-        self._publish("will_decode_instruction", self.PC)
+        self._publish("will_decode_instruction", curpc)
```

Comment:
I suppose `will_decode_instruction` can also change the state, so we need to "invalidate" the cached value again

Reply:
Is there a use case for this ?
Or is this a general principle when publishing an event ?

Reply:
I think we should not rely that somebody won't change any state during any event.

Regarding use case... What event is used when `@hook` on an address is used?

If that is not `will_decode_instruction` and is `will_execute_instruction` then maybe it should be the first one :thinking:. Or at least someone might use that as an optimization.

Reply:
For `@hook`,  it is `will_execute_instruction`

After thinking, I agree that wa cannot assume change state during event `will_decode_instruction` So, I refresh PC value once.

Optimization is a bit less effective, but still good to have ;-)